### PR TITLE
[CI] Add Release Notes Automation

### PR DESCRIPTION
**Pull request by jackkoenig**
Add new workflow to generate release notes. This workflow can be run manually via workflow dispatch, but will also run automatically upon creating a release. The workflow dispatch can be used for testing, previewing release notes for a potential release, or for generating release notes for past releases for when the automation has a bug or didn't exist.

### Contributor Checklist

- [NA] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Internal or build-related (includes code refactoring/cleanup)

#### API Impact

No impact

#### Backend Code Generation Impact

No impact

#### Desired Merge Strategy

- Squash

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.5.x` or `3.6.x` depending on impact, API modification or big change: `5.0.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by jackkoenig** [comment]
The new `Require Release Notes Label` correctly blocked this PR and then took ~10s to update once I added a label.

File: .github/workflows/release-notes.yml
Comment (jackkoenig) [inline .github/workflows/release-notes.yml:33]:
```suggestion
```
Delete unneeded reference to this repo

File: .github/workflows/release-notes.yml
Comment (jackkoenig) [inline .github/workflows/release-notes.yml:30]:
```suggestion
          configuration: .github/workflows/config/release-notes.json
          failOnError: true
```
Enable failing on error, default is to just ignore errors and thus get an empty output.

**Review by chick — APPROVED** [review]
Looks good, didn't dig too deep. Let me know if I should do a deeper dive

**Review by mwachs5 — COMMENTED** [review]
is there anythign to add to CONTRIBUTING.md or the PR checklist that goes in hand with this?

**Comment by jackkoenig** [comment]
> is there anything to add to CONTRIBUTING.md or the PR checklist that goes in hand with this?

Yeah probably, I'll add a couple of things. We can also iterate as the rest of the developers (who missed today's dev meeting where I talked about this 😉) become aware of the new automation and rules.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (mwachs5) [inline .github/PULL_REQUEST_TEMPLATE.md:16]: can we just take this off and tell people to select the right label? So they don't have to mess with the PR template AND apply the label?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (mwachs5) [inline .github/PULL_REQUEST_TEMPLATE.md:15]:
```suggestion
<!-- Choose one or more from the following, by applying the correct label(s)
```

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (jackkoenig) [reply .github/PULL_REQUEST_TEMPLATE.md:16]: You must have write access to set the label, so unfortunately no.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (mwachs5) [inline .github/PULL_REQUEST_TEMPLATE.md:26]: should these be labels too? Or say something like "if you picked API Modification above, explain more here"?

Comment (mwachs5) [reply .github/PULL_REQUEST_TEMPLATE.md:26]: same with below "Backend Code Generation Impact" -- should that be a label?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (jackkoenig) [reply .github/PULL_REQUEST_TEMPLATE.md:16]: We could make a workflow that auto labels based on the body of the PR though, but the usefulness vs. amount of work relative to how easy it is to label a PR makes me feel it isn't worth it... yet at least.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (mwachs5) [reply .github/PULL_REQUEST_TEMPLATE.md:26]: And can we get rid of "desired merge strategy"...? 

Mostly trying to slash the words people write that are redundant with labels they will apply.

Mayve this is getting too demanding for this PR :)

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (jackkoenig) [reply .github/PULL_REQUEST_TEMPLATE.md:15]: Those without write access cannot label the PR, so I put the onus of applying the correct label on the reviewer down below.

Comment (mwachs5) [reply .github/PULL_REQUEST_TEMPLATE.md:15]: got it.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (mwachs5) [inline .github/PULL_REQUEST_TEMPLATE.md:49]: does this have to be one?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (jackkoenig) [reply .github/PULL_REQUEST_TEMPLATE.md:26]: > And can we get rid of "desired merge strategy"...?
> 
> Mostly trying to slash the words people write that are redundant with labels they will apply.
> 
> Mayve this is getting too demanding for this PR :)

I'm all for some streamlining of the template, but I would rather save that for a separate PR

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (mwachs5) [reply .github/PULL_REQUEST_TEMPLATE.md:49]: (i agree it's probably good practice sort them by order of precedence if it can be only one)

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (jackkoenig) [reply .github/PULL_REQUEST_TEMPLATE.md:26]: Actually I am going to tweak it a bit because things like API and Backend Code Generation Impact should be included in release notes and thus shouldn't be separate sections.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (jackkoenig) [reply .github/PULL_REQUEST_TEMPLATE.md:45]: image

Comment (jackkoenig) [reply .github/PULL_REQUEST_TEMPLATE.md:45]: Since each PR can only be in one category in the release notes (I guess we could change that but IMO it only makes sense to be there once), there can only be one release notes label. We could maybe change this but I think it's a reasonable place to start and see how it goes.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (jackkoenig) [inline .github/PULL_REQUEST_TEMPLATE.md:30]: @mwachs5 I've made some more changes to the template to streamline it a little and nudge PR authors more toward writing release notes than putting things in these sections (which are/were not included in the release notes).

**Review by mwachs5 — APPROVED** [review]

**Pull request merged**
